app/models/user.rb

```diff
@@ -0,0 +1,17 @@
+class User < ApplicationRecord
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i
+
+  validates :name, presence: true, length: {maximum: Settings.max_name}
+  validates :email, presence: true, length: {maximum: Settings.max_email},
+    format: {with: VALID_EMAIL_REGEX}, uniqueness: {case_sensitive: false}
+  validates :password, presence: true, length: {minimum: Settings.min_pass}
+
+  before_save :email_downcase
+
+  has_secure_password
```

Comment:
hàm này ở đâu ra vậy?

Reply:
Hàm has_secure_password là hàm có sẵn do Rails cung cấp để giúp cho việc mã hóa mật khẩu

Reply:
Rails cung cấp ở đâu vậy?

Reply:
Khi gọi has_secure_password thì sẽ load gem BCrypt.
Thêm thuộc tính chỉ đọc vào model called password
Xác nhận mật khẩu được điền vào trường password_confirmation

Reply:
Là 1 method từ ActiveModel::SecurePassword

Reply:
modlufe `ActiveModel::SecurePassword` cung cấp những gì?

Reply:
Có chức năng để lưu trữ an toàn tất cả các mật khẩu đã được mã hóa.
Có 2 module con: ActiveModel::SecurePassword::ClassMethods và ActiveModel::SecurePassword::InstanceMethodsOnActivation
ActiveModel::SecurePassword::ClassMethods .Khi thêm phương thức has_secure_password thuộc module này, nó cung cấp để định nghĩa một mật khẩu xác nhận hợp lệ. .ActiveModel::SecurePassword phụ thuộc vào bcrypt(chức năng mã hoá mật khẩu), vì vậy để sử dụng model này một cách hiệu quả thì cần phải thêm gem bcrypt vào Gemfile. .Nó cung cấp một số chức năng như:
Mật khẩu phải tồn tại.
Mật khẩu phải khớp với phần xác nhận mật khẩu.
Mật khẩu không được vượt quá 72 ký tự.
ActiveModel::SecurePassword::InstanceMethodsOnActivation . Phương thức authenticate: Sẽ trả về self nếu mật khẩu đúng, nếu sai thì sẽ trả về false. . Phương thức password: Sẽ mã hóa mật khẩu khi được nhập, và chỉ thực hiện khi mật khẩu không được rỗng . Phương thức password_confirmation: Thực hiện chức năng xác nhận lại mật khẩu.